feat(settings): add an option to toggle identicons

tox-user opened the pull request:
Solves #4741

Adds a setting to User Interface tab that allows to toggle displaying
identicons instead of default avatar picture.

This toggles identicons for all of user's contacts, not just for the 
user's own profile.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/4818)
<!-- Reviewable:end -->

tox-user committed:
feat(settings): add an option to toggle identicons

Solves #4741

Adds a setting to User Interface tab that allows to toggle displaying
identicons instead of default avatar picture.

Diadlo commented:
1) Please, add information box or somewere information, that change will be accepted after restart

2) Issue:
2.1) enable identicon
2.2) restart qTox
2.3) disable identicon
2.4) restart qTox
2.5) Observed: you avatar not reseted

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/persistence/profile.cpp, line 398 at r1](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzVJYN-Y5dUIBEq41pj:-KzVJYN-Y5dUIBEq41pk:begm0k3) ([raw file](https://github.com/qtox/qtox/blob/9011a21c420ab4429ca603a92363fa6664177c98/src/persistence/profile.cpp#L398)):*
> ```cpp
> {
>     QPixmap pic;
>     if (Settings::getInstance().getShowIdenticons() == true) {
> ```

You don't need to compare bool with `true`. Just use
`if (Settings::getInstance().getShowIdenticons()) {`

---

*[src/persistence/profile.cpp, line 401 at r1](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzVJkeln7D4-RpYw5cx:-KzVJkeln7D4-RpYw5cy:biyh71h) ([raw file](https://github.com/qtox/qtox/blob/9011a21c420ab4429ca603a92363fa6664177c98/src/persistence/profile.cpp#L401)):*
> ```cpp
> 
>         const QByteArray avataData = loadAvatarData(owner);
>         if(avataData.isEmpty()) {
> ```

Add space after `if`

---

*[src/persistence/profile.cpp, line 406 at r1](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzVJxG_S9vd1mTSzNI6:-KzVJxG_S9vd1mTSzNI7:bvt1fsf) ([raw file](https://github.com/qtox/qtox/blob/9011a21c420ab4429ca603a92363fa6664177c98/src/persistence/profile.cpp#L406)):*
> ```cpp
>             pic.loadFromData(avataData);
>         }
>     
> ```

Seems tab here (or reviewable strange symbols). If there is tab, please, remove it. If not -- ignore

---

*[src/persistence/profile.cpp, line 476 at r1](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzVJt2AKO6NoHnTNYGD:-KzVJt2AKO6NoHnTNYGE:b49haa5) ([raw file](https://github.com/qtox/qtox/blob/9011a21c420ab4429ca603a92363fa6664177c98/src/persistence/profile.cpp#L476)):*
> ```cpp
>         avatarData = pic;
>     } else {
>         if (Settings::getInstance().getShowIdenticons() == true) {
> ```

Same with `== true`

---

*[src/persistence/profile.cpp, line 483 at r1](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzVJvpRMDqhU6lWUJxE:-KzVJvpRMDqhU6lWUJxF:b7x1i6y) ([raw file](https://github.com/qtox/qtox/blob/9011a21c420ab4429ca603a92363fa6664177c98/src/persistence/profile.cpp#L483)):*
> ```cpp
>             buf.open(QIODevice::WriteOnly);
>             identicon.save(&buf, "png");
>             
> ```

Same with tab

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4818#-:-KzVMyMvR5aRypXK3am5:b49y07b)*
<!-- Sent from Reviewable.io -->

tox-user commented:
1. That information is already in the tootip (https://github.com/qTox/qTox/pull/4818/commits/905ca7708662c6fe5520ad2707715075ad1aa41d#diff-09d87ecacfd6805618183e4648cec68dR263).
2. This is because avatars are saved to a file. How do I distinguish if a file is a user avatar or an identicon? I'm stuck.

---

Review status: 5 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/persistence/profile.cpp, line 398 at r1](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzVJYN-Y5dUIBEq41pj:-KzVaqtCrUfP4nR-m6HM:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/9011a21c420ab4429ca603a92363fa6664177c98/src/persistence/profile.cpp#L398)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

You don't need to compare bool with `true`. Just use
`if (Settings::getInstance().getShowIdenticons()) {`
</blockquote></details>

Done.

---

*[src/persistence/profile.cpp, line 401 at r1](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzVJkeln7D4-RpYw5cx:-KzVarfC2hdWME0KEewb:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/9011a21c420ab4429ca603a92363fa6664177c98/src/persistence/profile.cpp#L401)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Add space after `if`
</blockquote></details>

Done.

---

*[src/persistence/profile.cpp, line 406 at r1](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzVJxG_S9vd1mTSzNI6:-KzVaZPO_Px650ns6rHm:b-1chgip) ([raw file](https://github.com/qtox/qtox/blob/9011a21c420ab4429ca603a92363fa6664177c98/src/persistence/profile.cpp#L406)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Seems tab here (or reviewable strange symbols). If there is tab, please, remove it. If not -- ignore
</blockquote></details>

There is no tab there.

---

*[src/persistence/profile.cpp, line 476 at r1](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzVJt2AKO6NoHnTNYGD:-KzVauFp9fu2Y_Rk4J3m:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/9011a21c420ab4429ca603a92363fa6664177c98/src/persistence/profile.cpp#L476)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Same with `== true`
</blockquote></details>

Done.

---

*[src/persistence/profile.cpp, line 483 at r1](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzVJvpRMDqhU6lWUJxE:-KzVahyte89Q_3uuI6Su:b-ci1cfc) ([raw file](https://github.com/qtox/qtox/blob/9011a21c420ab4429ca603a92363fa6664177c98/src/persistence/profile.cpp#L483)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Same with tab
</blockquote></details>

There is no tab there either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4818#-:-KzVb-n_eUCK4KJ4lRe0:b-cw7ley)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
@sudden6 Probably, qTox should show identicon instead of default avatar, but not set it?

Diadlo commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4818#-:-KzXQ1yHbB3Jr_o5_IgX:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

tox-user commented:
@Diadlo perhaps there should be one more option for user's own avatar? It would allow to switch between identicon and default picture for user's profile.

So there would be two independent options:
▢ Use identicons for contacts (when their avatar is empty)
▢ Use identicon for my own profile (when my avatar is empty)

I still don't know how it would work though.

Diadlo commented:
@tox-user I don't think, that the second point make sense in practice. If someone want to see identicon in one place, he wants in another places too.
But qTox shouldn't set identicon as avatar, just show it.

Diadlo commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[src/persistence/profile.cpp, line 406 at r2](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzZMUTZowDe6Gom3U2m:-KzZMUT_ACv6jcesnT4P:b-nekgxc) ([raw file](https://github.com/qtox/qtox/blob/905ca7708662c6fe5520ad2707715075ad1aa41d/src/persistence/profile.cpp#L406)):*
> ```cpp
>             pic.loadFromData(avataData);
>         }
>     
> ```

Oh.. it's spaces. Please, remove empty line here or spaces in the begin of the line

---

*[src/persistence/profile.cpp, line 483 at r2](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzZM_YeBAH55Kt5bEY0:-KzZM_YeBAH55Kt5bEY1:b-ya81ng) ([raw file](https://github.com/qtox/qtox/blob/905ca7708662c6fe5520ad2707715075ad1aa41d/src/persistence/profile.cpp#L483)):*
> ```cpp
>             buf.open(QIODevice::WriteOnly);
>             identicon.save(&buf, "png");
>             
> ```

Same as above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4818)*
<!-- Sent from Reviewable.io -->

sudden6 commented:
@Diadlo IMO if it's shown to the user it should also be sent to others as avatar, less confusion possibilities.

tox-user commented:
@sudden6 so according to what you say if I have no avatar set and I have identicons enabled, someone who has them disabled will still see my identicon, because it has been set as my avatar. The setting will not work properly then. It will fail to disable identicons.

Right now it can't distinguish an avatar from an identicon. When user creates a new profile with no avatar, qTox will create an indenticon for them and set it as their avatar. This is because identicons are enabled by default. When the same user later decides they don't want to see identicons anymore and they disable them, their profile will still have an identicon as avatar.

This is all caused by the fact that our own identicon is saved to a file like any other avatar, so we can never be sure if user's avatar has actually been set or not, because if it hasn't qTox will set one for them. That means this PR in its current state doesn't work and we need a new approach.

Diadlo commented:
@sudden6 When we show default avatar, but not set, it doesn't confuse people, isn't it?

tox-user commented:
@Diadlo for users with no avatar who are on older versions of qTox or who use other clients we see their identicons even though their clients don't have that feature. So we already see something that is not there.

Diadlo commented:
@tox-user I know. I talk about @sudden6 argument with confusing. We didn't set qTox default avatar image as avatar before, we just showed it

tox-user committed:
fix(identicon): don't set identicon as an avatar

Do not save user's identicon to a file like a normal avatar. This allows
'Use identicons' setting to affect user's own identicon.

tox-user commented:
I removed the part that saved the identicon as avatar. It solved the problem. If we want to be able to toggle identicons I don't see any other way.

---

Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[src/persistence/profile.cpp, line 406 at r2](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzZMUTZowDe6Gom3U2m:-Kz_mUCPMj3d9z37I8DN:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/905ca7708662c6fe5520ad2707715075ad1aa41d/src/persistence/profile.cpp#L406)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Oh.. it's spaces. Please, remove empty line here or spaces in the begin of the line
</blockquote></details>

Done.

---

*[src/persistence/profile.cpp, line 483 at r2](https://reviewable.io:443/reviews/qtox/qtox/4818#-KzZM_YeBAH55Kt5bEY0:-Kz_mUmup420ytcINI3L:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/905ca7708662c6fe5520ad2707715075ad1aa41d/src/persistence/profile.cpp#L483)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Same as above
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4818#-:-Kz_mZYaxwQEd4MfzhiG:b-pprg0w)*
<!-- Sent from Reviewable.io -->

Diadlo commented:
@sudden6 So what is the decision? Two options in settings (I'm against this way) or not set identicon as avatar? Or set, when identicon enabled and remove, when disabled?

anthonybilinski commented:
I think it makes sense to treat identicons like the default avatar picture, where it's displayed to the user but not sent to different contacts like a real avatar picture is. It's true it would cause a different image to display for a user and their friend, but if their friend's client implements the identicon spec, they will see the same identicon if their client supports it and they have it enabled.

Note that there are some backwards compatibility issues here. Even if we stop saving identicons as avatar pictures, for any user that has run qTox between when identicons were added until when this gets merged, they will still have those identicons saved locally and will still load them as avatars even if they disable identicons in settings. This would affect any user that has used 0.12.*.

I'm not sure if that's worth a work-around, that could be when the setting is disabled by a user, we check if they already have the identicon image saved locally and if so delete it?

Diadlo commented:
@anthonybilinski qTox with identicon implementation wasn't released yet. Masten is unstable and its use is carried out at your own risk. We are not required to comply with backward compatibility with the master.

So, I think we should just remove setting identicon before release and everything will be fine

anthonybilinski commented:
Good point! I thought it was added right before 0.12, but it was really added right after.

tox-user commented:
> It's true it would cause a different image to display for a user and their friend, but if their friend's client implements the identicon spec, they will see the same identicon if their client supports it and they have it enabled.

Identicons don't make sense, if all clients don't use the same spec anyway.

> I'm not sure if that's worth a work-around, that could be when the setting is disabled by a user, we  check if they already have the identicon image saved locally and if so delete it?

@anthonybilinski, @Diadlo is kind of right about compatibility, but I think we should try not to break it if possible. The problem is how do we know for sure, that the image we want to delete is an identicon and not user's avatar? We don't want to start deleting people's avatars and I don't see how we could distinguish them. If we could somehow do that then we wouldn't have the compatibility problem anyway, because we could just keep saving them as avatars and delete them whenever users disable them.

anthonybilinski commented:
@tox-user  We can generate the identicon and check if it's binary equivalent to the saved PNG, and if so assume it was an auto-generated avatar that the user didn't set themselves. The user could still theoretically manually set their avatar to the PNG that is generated by the identicon algo, but that's pretty crazy.

sudden6 commented:
@Diadlo I understand now what you mean and agree with you and @tox-user that it's better not to save the Identicon as a file.

@anthonybilinski This could also be a good idea

sudden6 commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 5 of 6 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4818#-:-Kzk36alhhwhft55eUp-:b-jbsf8z)*
<!-- Sent from Reviewable.io -->
